src/main/java/com/bisnode/opa/process/OpaTestProcess.java

```diff
@@ -24,16 +26,28 @@ public ProcessExecutionResult execute() {
                     .directory(rootDir)
                     .command(command.getCommandArgs())
                     .start();
+
+            OpaOutputConsumer opaOutputConsumer = new OpaOutputConsumer(process);
+            String testResultFromOpa = compileOpaOutput(opaOutputConsumer);
             int exitCode = process.waitFor();
-            return new ProcessExecutionResult(asString(process.getInputStream()), exitCode);
+            // Should only retrieve results after OPA process exits
+            return new ProcessExecutionResult(testResultFromOpa, exitCode);
         } catch (IOException | InterruptedException e) {
             throw new TestExecutionException("Failed to start OPA process for tests", e);
         }
     }
 
-    private String asString(InputStream inputStream) {
-        // "Stupid Scanner trick" https://community.oracle.com/blogs/pat/2004/10/23/stupid-scanner-tricks
-        return new Scanner(inputStream, UTF_8.name()).useDelimiter("\\A").next();
+    private String compileOpaOutput(OpaOutputConsumer opaOutputConsumer) {
```

Comment:
I think this should be a part of `OpaOutputConsumer`, what's your take @k-jingyang?

Reply:
Your question gave me a new idea on an improvement. :smile: